### DISCUSSION TITLE
Add a type utility for getting all events (including internal ones) from a state machine

### DISCUSSION THREAD

**Discussion by tom-sherman** [statelyai/xstate#3117] [Ideas]
Some discussion on Discord about this: https://discord.com/channels/795785288994652170/809564589880639548/948993351431118888

I want a union of all of my external and internal events.

Right now I have to do this:

```ts
type AllEvents =
  | MyMachine['__TResolvedTypesMeta']['internalEvents'][keyof MyMachine['__TResolvedTypesMeta']['internalEvents']]
  | MyMachineEvents
```

**Comment by Andarist** [comment]
The easiest way to do this would be something like:
```ts
type Values<T> = T[keyof T]
type AllMachineEvents = Values< MyMachine["__TResolvedTypesMeta"]["indexedEvents"]>
```

I wonder though, why do you need this? The Discord thread doesn't mention this.

**Reply by tom-sherman** [reply]
Thanks for the implementation tip, will go ahead with that as it's much easier to read.

> I wonder though, why do you need this?

I have some utlities that sit on `onTransition` that act on the events. One example of these utilities is logging, we use the types to ensure we strip out customer information from events.

**Reply by Andarist** [reply]
Do you have ane name for such a type? The current `EventFrom` is assuming that it should only return events that can be sent by the user and I think that's a good default. Since this type already accepts some extra generic I think that it makes sense to create a separate type for this and not add more "params" to this one. :thinking:

**Reply by tom-sherman** [reply]
Makes sense for it to be a separate type. I don't mind David's suggestion of `OmniEvent`, although that might be a bit esoteric.

You could avoid the naming problem and add a new type for `InternalEventFrom`. Then I can do `InternalEventFrom<MyMachine> | EventFrom<MyMachine>` without relying on the structure of `__TResolvedTypesMeta`.

**Reply by Andarist** [reply]
> You could avoid the naming problem and add a new type for InternalEventFrom. Then I can do InternalEventFrom<MyMachine> | EventFrom<MyMachine> without relying on the structure of __TResolvedTypesMeta.

That might be a good idea - I'd have to think how to best structure this, cause `MyMachine["__TResolvedTypesMeta"]["internalEvents"]` isn't everything that constitutes for "internal events", see the complex-ish logic for creating final `indexedEvents` here:
https://github.com/statelyai/xstate/blob/28e3530818e1d800eba7b6d821bde0c0048f0579/packages/core/src/typegenTypes.ts#L175-L184

I guess the minimal version that wouldn't go into any optimizations would be just this:
```ts
type Cast<A, B> = A extends B ? A : B;
type Values<T> = T[keyof T];

type InternalEvents<
  TMachine extends AnyStateMachine,
  TIndexedEvents = TMachine["__TResolvedTypesMeta"]["indexedEvents"],
  TPublicType = TMachine["__TEvent"]["type"]
> = Values<
  string extends TPublicType
    ? TIndexedEvents
    : Omit<TIndexedEvents, Cast<TPublicType, string>>
>;
```

**Reply by yitaya1** [reply]
Hey there, since `__TResolvedTypesMeta` is marked as `@deprecated`, does it mean that future implementation of this `OmniEvent` is subject to change?

In my use case I wish to create a map of `eventName, eventType`.

**Reply by tom-sherman** [reply]
I think it's marked as deprecated mostly so that it doesn't appear in autocompletion and such, I think it is supposed to be internal and so ideally shouldn't be relied upon.

**Reply by Andarist** [reply]
Yeah, it's exactly like @tom-sherman has said. The shape of the types meta is subject to change, so we can't allow people to interact with it directly.

> In my use case I wish to create a map of eventName, eventType.

Creating a map like this can be somewhat easily done with a helper like this:
```ts
type IndexByType<T extends { type: string }> = {
  [E in T as E["type"]]: E
};
```
But, of course, you still need to supply some union to this as a type argument.

I've created [a draft PR](https://github.com/statelyai/xstate/pull/3160) with the implementation that I've used here for `InternalEvents`, but it's still a subject of discussion. I'm eager to hear out your thoughts there.

**Reply by yitaya1** [reply]
I see, thanks for clarifying the status of `__TResolvedTypesMeta`.

As for the events I ended up using:
```
declare type Dictionary<T> = keyof T extends infer K // introduce type parameter
  ? K extends keyof T // distribute over the type parameter
    ? { key: K; value: T[K] } // The type that we construct for each K
    : never
  : never;

type EventNameAndTypeMapByMachine<T extends AnyStateMachine> = Dictionary<
  Omit<
    T['__TResolvedTypesMeta']['indexedEvents'],
    keyof T['__TResolvedTypesMeta']['internalEvents']
  >
>;
```

Which yields a generic map for a machine's events.

**Reply by Andarist** [reply]
I really don't think that you should depend on `T['__TResolvedTypesMeta']['internalEvents']`, see the reasoning [here](https://github.com/statelyai/xstate/discussions/3117#discussioncomment-2295224). 

I also wonder - why do u need to create such a union? It's a different one that was originally requested in this thread (it boxes events in an object like `{ k: K, value: T[K] }`).

And note that the shape of `T['__TResolvedTypesMeta']` has been recently changed in: https://github.com/statelyai/xstate/pull/3153 . This is why you should never rely on it - it's not part of the public API.
